### ISSUE TITLE
Feature: multiple table schemas generated in one file

### ISSUE THREAD

**Issue by ryantm** [khibino/haskell-relational-record#46]
In your 2015 whitepaper, section 4.1, you mention that OverloadedRecordFields would enable the feature of generating multiple table schemas in the same file. Has any work been done on this? I am interested in working on it if there hasn't been much work done yet. Any suggestions or hints about how to do it would be appreciated.

**Comment by kazu-yamamoto** [comment]
OverloadedRecordFields has sad history. It was divided into three extensions and two of them are now in GHC 8.0. If I understand correctly, it's incomplete and does not work as what we expect. We need to wait for the progression. https://ghc.haskell.org/trac/ghc/wiki/Records/OverloadedRecordFields

**Comment by ryantm** [comment]
Since we are talking about using Template Haskell, it seems like the DuplicateRecordFields extension is enough to generate the entire schema in one file if the Pi projections are given unique names. It seems that Magic type classes would allow us to stop qualifying or using unique names for the Pi projections, but it seems like we could implement our own specific version of them, like the lens library does.

**Comment by ryantm** [comment]
Generating something like this would allow us to avoid qualifying or making the Pi projections have unique names: https://gist.github.com/ryantm/b77d041d55d4be015abbc3036844def9

**Comment by kazu-yamamoto** [comment]
Good point. Yes, we are using template haskell.
Though written in Japanese, please give a look at this article: http://d.hatena.ne.jp/kazu-yamamoto/20160114/1452735514
You can see how we should use OverloadedLabels.

**Comment by ryantm** [comment]
I started working on this feature. I have a working and badly in need of refactoring version of the code up at https://github.com/ryantm/haskell-relational-record master that lets you use mutiple `defineTableFromDB'` splices in the same module if you set the `NameConfig` called `overloadedRecords` to `True`.

Grabbing an example from the documentation and converting it would go from this:

```Haskell
employee_3_7_3 :: Relation () Employee1
employee_3_7_3 = relation $ do
  e <- query employee
  asc $ e ! Employee.title'
  asc $ e ! Employee.lname'
  return $ Employee1 |$| e ! Employee.empId'
                     |*| e ! Employee.title'
                     |*| e ! Employee.startDate'
                     |*| e ! Employee.fname'
                     |*| e ! Employee.lname'
```
To this:
```Haskell
employee_3_7_3 :: Relation () Employee1
employee_3_7_3 = relation $ do
  e <- query employee
  asc $ #title e
  asc $ #lname e
  return $ Employee1 |$| #empId e
                     |*| #title e
                     |*| #startDate e
                     |*| #fname e
                     |*| #lname e
```

I think this seems like a nice improvement to writing and somewhat to reading the code, and I'd be happy to hear feedback about it. I am not exactly happy with how the label comes before the table relation, but that's typically how Haskell works, so maybe it is okay.

I have a question for the Haskell Relational Record experts: Is there a useful difference between having `Relation () (Maybe Int)` vs `Relation () (Maybe (Maybe Int)`? The latter can arise when you are doing left joins, because a queryMaybe adds one Maybe and a Nullable field adds another. I am aware you can easily collapse the maybe with the `?!?` projection operator, but is it useful to make the distinction between `?!?` and `?!` ?

I thought maybe the useful difference would be you could distinguish between the join not being there and the column being Null, but I don't think you can even do that within a SQL query.

**Comment by khibino** [comment]
Thanks for your working about this feature.
Your implementation is very helpful to me to understand that OverloadedLabels extension is useful in this case.

I uploaded relational-query-0.11.0.0 to support this feature.

I defined both function label `Record c a -> Record c b` and projection label `Pi a b`,
so both function application style and field selector style are available. ( `#foo a` , `a ! #foo` )

Except for special cases,
It is impossible to determine a SQL NULL value is `Nothing` or `Just Nothing`.
So, I intend to avoid types like `Relation (Maybe (Maybe Int))`.

Maybe features as before are easily derived from projection labels like (? #foo) , (?? #foo).
( ? is same as ?!, and ?? is same as ?!?. )
